examples/camera-app/camera-common/src/video-source-impl.cpp

```diff
@@ -0,0 +1,320 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "video-source-impl.h"
+
+#include <gst/app/gstappsink.h>
+#include <gst/gst.h>
+#include <iostream>
+
+#include <chrono>
+#include <ctime>
+#include <iomanip>
+
+VideoSourceImpl::VideoSourceImpl()
+{
+    pipeline   = nullptr;
+    start_time = std::chrono::system_clock::now();
+}
+
+// Implement InitializeCameraDevice function
+CameraError VideoSourceImpl::InitializeCameraDevice()
+{
+    // Implement the initialization logic for the camera hardware
+    // Return CameraError::SUCCESS if initialization is successful, otherwise return an appropriate error code
+    return CameraError::SUCCESS;
+}
+
+// Implement InitializeStreams function
+CameraError VideoSourceImpl::InitializeStreams()
+{
+    // Implement the initialization logic for the camera streams
+    // Return CameraError::SUCCESS if initialization is successful, otherwise return an appropriate error code
+    return CameraError::SUCCESS;
+}
+
+// Callback function for processing frames
+GstFlowReturn VideoSourceImpl::on_new_sample(GstAppSink * appsink, gpointer user_data)
+{
+    GstSample * sample = gst_app_sink_pull_sample(appsink);
+    if (sample)
+    {
+        GstBuffer * buffer = gst_sample_get_buffer(sample);
+        GstMapInfo map;
+        if (gst_buffer_map(buffer, &map, GST_MAP_READ))
+        {
+            static int frame_count = 0;
+            ++frame_count;
+            std::time_t now_c = std::chrono::system_clock::to_time_t(start_time);
+            std::tm * now_tm  = std::localtime(&now_c);
+            std::cout << std::put_time(now_tm, "%Y-%m-%d %H:%M:%S") << " Frame: " << frame_count
+                      << " Received H.264 frame, size: " << map.size << " bytes" << std::endl;
+            // Send video buffers to transport
+            // video_frame_callback(map.data, map.size);
+            gst_buffer_unmap(buffer, &map);
+        }
+        else
+        {
+            std::cerr << "Failed to map buffer" << std::endl;
+        }
+        gst_sample_unref(sample);
+        return GST_FLOW_OK;
+    }
+    else
+    {
+        std::cerr << "Failed to pull sample" << std::endl;
+        return GST_FLOW_ERROR;
+    }
+}
+
+// Implement VideoStreamAllocate function
+CameraError VideoSourceImpl::VideoStreamAllocate(const VideoStreamStruct & videoStreamParams, uint16_t & outStreamID)
+{
+
+    video_stream_params = videoStreamParams;
+
+    // Create pipeline
+    pipeline = (GstElement *) gst_pipeline_new("video-pipeline");
+    if (!pipeline)
+    {
+        std::cerr << "Failed to create pipeline!" << std::endl;
+        return -1;
+    }
+
+    // Create elements
+    GstElement * videotestsrc = gst_element_factory_make("videotestsrc", "source");
```

Comment:
If using videotestsource then I think it might be useful to bubble this up as a command line parameter to the app so that we can specify whether to use a real video source vs a test source.